contracts/ROSCA.sol

```diff
@@ -0,0 +1,222 @@
+      throw;
+
+    if (currentRound != 0) {
+      if (winnerAddress == 0) //only true when there is no bidder in this round
+      {
+        // there is no bid in this round so find an unpaid address for this epoch
+        for (uint i = 0 ; i < membersAddresses.length; i++)
+        {
+          uint semi_random = now % membersAddresses.length;
+          if(members[membersAddresses[(currentRound-1 + i) % membersAddresses.length]].paid < currentEpoch)
+          winnerAddress = membersAddresses[currentRound-1 + i];
+          break;
+        }
+      }
+      if(members[winnerAddress].paid == currentEpoch) throw;
+      members[winnerAddress].pendingWithdrawl += lowestBid;
+      members[winnerAddress].paid++;
+      LogRoundFundsReleased(winnerAddress, lowestBid);
+    }
+    if (currentRound < membersAddresses.length)  // reset variables related to bidding
+    {
+      lowestBid = contributionSize * membersAddresses.length;
+      winnerAddress = 0;
+
+      currentRound++;
+    }else if (currentEpoch < numEpochs) // end of Epoch reset rounds
+    {
+      currentEpoch++;
+      currentRound = 1;
+      lowestBid = contributionSize;
+    }
+  }
+
+  /**
+   * Anyone not already a member of ROSCA can request to join and they'll be put into
+   * pendingJoinRequest untill foreman accept request or ROSCA has started
+   *
+   */
+  function joinRequest() onlyBeforeStart
+  {
+    // only put the request in the pending list if they are not in the ROSCA already
+    if (members[msg.sender].alive) throw;
+    pendingJoinRequest[msg.sender] = true;
+    LogParticipantApplied(msg.sender);
+  }
+
+  /**
+   *
+   * once a requestor had been registered as member, the address will be taken out of pendingJoinRequest
+   **/
+  function acceptJoinRequest(address requestor)
+    onlyForeman
+    beforeStart
+  {
+    if (!pendingJoinRequest[requestor]) throw;
+    members[requestor] = USER({paid: 0 , contributed: 0, alive: true, pendingWithdrawl: 0});
+    membersAddresses.push(requestor);
+    numRoundsInEpoch++;
+    LogParticipantApproved(requestor);
+    delete(pendingJoinRequest[requestor]); // take out the requestor's address in the pending list
+  }
+
+  /**
+   * Processes a periodic contribution from msg.sender ().
+   * Any excess funds will be withdrawable through withdraw().
+   */
+  function contribute() payable
+  {
+    if (members[msg.sender].alive)
+    {
+      members[msg.sender].contributed += msg.value;
+      LogContributionMade(msg.sender, msg.value);
+    }
+  }
+
+  /**
+   * Registers a bid from msg.sender. If msg.sender has already won a round or bid is higher than lowestBid,
+   * this method will throw.
+   */
+  function bid(uint distrubtionAmountInWei)
+  {
+    if (!(distrubtionAmountInWei < lowestBid && members[msg.sender].paid < currentEpoch && currentRound > 0
```

Comment:
- Instead of if (!(a > x && b < y && c > z)) consider using if (a <= x || b >= y || c <= z). It might be easier to read.
One way or another, this is pretty complex logic, and our penalty for mistakes is huge (users losing money, we lost reputation). Consider breaking this into a few simpler lines.